### PR TITLE
Use flag for decompression error

### PR DESCRIPTION
**Pull request by anikamukherji**
Description: Add flag for decompression failure rather than hard assert failure. This is a patch that Pinterest has been running for several months now.

Risk Level: Low
Testing: Unit tests
Docs Changes: N/A
Release Notes: Added

Signed-off-by: Anika Mukherji <amukherji@pinterest.com>

### CODE REVIEW

File: source/common/decompressor/zlib_decompressor_impl.cc [hunk 1]
```diff
@@ -68,7 +68,11 @@ bool ZlibDecompressorImpl::inflateNext() {
     return false; // This means that zlib needs more input, so stop here.
   }
 
-  RELEASE_ASSERT(result == Z_OK, "");
+  if (result < 0) {
+    decompression_error_ = result;
```
Comment (jmarantz) [inline source/common/decompressor/zlib_decompressor_impl.cc]: suggest adding trace-log here with the error number & URL.

File: test/common/decompressor/zlib_decompressor_impl_test.cc [hunk 2]
```diff
@@ -70,13 +71,15 @@ class ZlibDecompressorImplDeathTest : public ZlibDecompressorImplTest {
...
 };
 
-// Exercises death by passing bad initialization params or by calling decompress before init.
-TEST_F(ZlibDecompressorImplDeathTest, DecompressorDeathTest) {
+// Test different failures by passing bad initialization params or by calling decompress before
+// init.
+TEST_F(ZlibDecompressorImplFailureTest, DecompressorFailureTest) {
   EXPECT_DEATH_LOG_TO_STDERR(decompressorBadInitTestHelper(100), "assert failure: result >= 0");
-  EXPECT_DEATH_LOG_TO_STDERR(unitializedDecompressorTestHelper(), "assert failure: result == Z_OK");
+  unitializedDecompressorTestHelper();
```
Comment (dio) [inline test/common/decompressor/zlib_decompressor_impl_test.cc]: s/unitialized/uninitialized

File: source/common/decompressor/zlib_decompressor_impl.h [hunk 3]
```diff
@@ -43,6 +43,10 @@ class ZlibDecompressorImpl : public Decompressor {
   // Decompressor
   void decompress(const Buffer::Instance& input_buffer, Buffer::Instance& output_buffer) override;
 
+  // Int to track whether error occurred during decompression
```
Comment (dio) [inline source/common/decompressor/zlib_decompressor_impl.h]: nit: s/Int/An integer? Or A code?

File: source/common/decompressor/zlib_decompressor_impl.h [hunk 4]
```diff
@@ -43,6 +43,10 @@ class ZlibDecompressorImpl : public Decompressor {
...
+  // When an error occurs, the result (a negative int) will be stored in this variable
```
Comment (dio) [inline source/common/decompressor/zlib_decompressor_impl.h]: Sorry, super nit a "." at the end.

File: source/common/decompressor/zlib_decompressor_impl.cc (hunk 1 above)
Comment (anikamukherji) [reply source/common/decompressor/zlib_decompressor_impl.cc]: @jmarantz to which URL are you referring?

Comment (anikamukherji) [reply source/common/decompressor/zlib_decompressor_impl.cc]: I added a URL to the definitions of the error codes.

File: docs/root/intro/version_history.rst [hunk 5]
```diff
@@ -9,6 +9,7 @@ Version history
 * buffer: remove old implementation
 * build: official released binary is now built against libc++.
 * cluster: added :ref: `aggregate cluster <arch_overview_aggregate_cluster>` that allows load balancing between clusters.
+* decompressor: Remove decompressor hard assert failure and replace with an error flag.
```
Comment (dio) [inline docs/root/intro/version_history.rst]: Pardon me, s/Remove/remove following the style with other entries.

**Comment by derekargueta** [comment]
unrelated problem with Windows CI:
```
CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
```

**Comment by derekargueta** [comment]
(apologies for the force push, we had a DCO hiccup)

**Comment by anikamukherji** [comment]
Still seeing unrelated windows failure
`CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.`

**Comment by derekargueta** [comment]
@anikamukherji lizan is working on fixing that here: https://github.com/envoyproxy/envoy/pull/9305 it's not just your PR; once it's fixed you can merge master.

**Comment by lizan** [comment]
feel free to ignore windows CI they aren't required anyway.

**Pull request merged**